velox/functions/prestosql/DecimalArithmetic.cpp

```diff
@@ -0,0 +1,251 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/expression/DecodedArgs.h"
+#include "velox/expression/VectorFunction.h"
+#include "velox/functions/lib/LambdaFunctionUtil.h"
+#include "velox/type/DecimalUtils.h"
+
+namespace facebook::velox::functions {
+namespace {
+
+template <
+    typename R /* Result Type */,
+    typename A /* Argument1 */,
+    typename B /* Argument2 */,
+    typename Operation /* Arithmetic operation */>
+class DecimalBaseFunction : public exec::VectorFunction {
+ public:
+  DecimalBaseFunction(const uint8_t aRescale, const uint8_t bRescale)
+      : aRescale_(aRescale), bRescale_(bRescale) {}
+  void apply(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args,
+      const TypePtr& resultType,
+      exec::EvalCtx* context,
+      VectorPtr* result) const override {
+    if (args[0]->isConstantEncoding() && args[1]->isFlatEncoding()) {
+      // Fast path for (flat, const).
+      auto constant = args[0]->asUnchecked<SimpleVector<A>>()->valueAt(0);
+      auto flatValues = args[1]->asUnchecked<FlatVector<B>>();
+      auto rawValues = flatValues->mutableRawValues();
+      auto rawResults =
+          getRawResults(rows, args[1], resultType, context, result);
+      context->applyToSelectedNoThrow(rows, [&](auto row) {
+        Operation::template apply<R, A, B>(
+            rawResults[row], constant, rawValues[row], aRescale_, bRescale_);
+      });
+    } else if (args[0]->isFlatEncoding() && args[1]->isConstantEncoding()) {
+      auto flatValues = args[0]->asUnchecked<FlatVector<A>>();
+      auto constant = args[1]->asUnchecked<SimpleVector<B>>()->valueAt(0);
+      auto rawValues = flatValues->mutableRawValues();
+      auto rawResults =
+          getRawResults(rows, args[1], resultType, context, result);
+      context->applyToSelectedNoThrow(rows, [&](auto row) {
+        Operation::template apply<R, A, B>(
+            rawResults[row], rawValues[row], constant, aRescale_, bRescale_);
+      });
+    } else if (args[0]->isFlatEncoding() && args[1]->isFlatEncoding()) {
+      auto flatA = args[0]->asUnchecked<FlatVector<A>>();
+      auto rawA = flatA->mutableRawValues();
+      auto flatB = args[1]->asUnchecked<FlatVector<B>>();
+      auto rawB = flatB->mutableRawValues();
+
+      R* rawResults = nullptr;
+      if (!(*result)) {
```

Comment:
add a comment for this block that we are checking for reusing one of the input arguments for results

Reply:
Done.